Moving robot (#45)
ignitionrobotics/docs: tutorials/Moving_robot/moving_robot.md

# Moving the robot

> In this tutorial we will learn how to move our robot. We will use the robot we built in [Build your own robot](../SDF/car_demo.md) tutorial. You can download the robot from [here](../SDF/car_demo.sdf). You can also find the finished world of this tutorial [here](move_robot.sdf).

Reviewer: I think we should use the same file, with the same name, across all tutorials

> In this tutorial we will learn how to move our robot. We will use the robot we built in [Build your own robot](../SDF/car_demo.md) tutorial. You can download the robot from [here](../SDF/car_demo.sdf). You can also find the finished world of this tutorial [here](move_robot.sdf).

Author: In other tutorial I mentioned that we will rename the file. My motivation for doing so is have the final world file with the same name as the one used in the tutorials

> In this tutorial we will learn how to move our robot. We will use the robot we built in [Build your own robot](../SDF/car_demo.md) tutorial. You can download the robot from [here](../SDF/car_demo.sdf). You can also find the finished world of this tutorial [here](move_robot.sdf).

Reviewer: humm ok, I need to go over the tutorials again. I'll let you know if it's confusing when I do that, no need to change anything now.

# Moving the robot

> In this tutorial we will learn how to move our robot. We will use the robot we built in the [Build your own robot](../SDF/car_demo.md) tutorial. You can download the robot from [here](../SDF/car_demo.sdf). You can also find the finished world of this tutorial [here](moving_robot.sdf).

Reviewer: ```suggestion
In this tutorial we will learn how to move our robot. We will use the
robot we built in the [Build your own robot](../SDF/building_robot.md)
tutorial. You can download the robot from [here](../SDF/building_robot.sdf).
You can also find the finished world of this tutorial [here](moving_robot.sdf).
```

Author: Yes, I forget to push it

## What is a plugin

To make our robot move we will use the `diff_drive` plugin. But before doing so let's answer the question "What is a plugin?" A plugin is a chunk of code that is compiled as a shared library and inserted into the simulation. Plugins make us control many aspects of the simulation like world, models etc.

### Diff_drive plugin

> `diff_drive` plugin helps us control our robot, specifically a robot that can be differentially driven. Let's setup the plugin on our robot. Open the `car_demo.sdf` and add the following code within the `vehicle_blue` model tags.

Reviewer: ```suggestion
`diff_drive` plugin helps us control our robot, specifically a robot that
can be differentially driven. Let's setup the plugin on our robot. Open
the `building_robot.sdf` and add the following code within the `vehicle_blue`
model tags.
```

</plugin>
```

The `<plugin>` tag has two attributes, `filename` which takes the library file name and `name` which takes the name of the plugin.
In the `<left_joint>` and `<right_joint>` tags we define the joints which connect the left and the right wheel with the body of the robot, in our case `left_wheel_joint` and `right_wheel_joint`. `<wheel_separation>` takes the distance between the two wheels.
Our robot has its `left_wheel` at 0.6 m and the `right_wheel` at -0.6 m in y-axis with respect to the `chassis`, so the `wheel_separation` is 1.2 m.
`<wheel_radius>` takes the radius of the wheel which was defined in the `<radius>` tag under the wheel link.
`<odom_publish_frequency>` sets the frequency (in `Hz`) by which our car will accept the moving commands.
`cmd_vel` is the input `<topic>` to the `DiffDrive` plugin.

## Topics and Messages

Now our model is ready. We just need to send commands (messages) to it.
These messages will be published (sent) on the `cmd_vel` topic defined above.

A topic is simply a name for grouping a specific set of messages or a particular service.
Our model will subscribe (listen) to the messages sent on the `cmd_vel` topic.

Launch the car world:

> `ign gazebo car_demo.sdf`

Reviewer: ```suggestion
`ign gazebo building_robot.sdf`
```

In another terminal let's send a message to to our car:

`ign topic -t "/cmd_vel" -m ignition.msgs.Twist -p "linear: {x: 0.5}, angular: {z: 0.05}"`

Now you should have your robot moving in the simulation.
**Note:** Don't forget to press the play button in the simulation.

The command specifies the topic to publish to after the `-t` option.
After the `-m` we specify the message type.
Our robot expects messages of type `Twist` which consists of two components, `linear` and `angular`.
After the `-p` option we specify the content (value) of the message: linear speed `x: 0.5` and angular speed `z: 0.05`.

**Hint:** You can know what every topic option does using this command: `ign topic -h`

> For more information about `Topics` and `Messages` in ignition check the [Transport library tutorials](https://ignitionrobotics.org/api/transport/9.0/tutorials.html)

Reviewer: ```suggestion
For more information about `Topics` and `Messages` in Ignition check the [Transport library tutorials](https://ignitionrobotics.org/api/transport/9.0/tutorials.html)
```

## Moving the robot using the keyboard

Instead of sending messages from the terminal we will send messages using the keyboard keys. To do so we will add two new plugins: `KeyPublisher` and `TriggeredPublisher`.

### KeyPublisher

`KeyPublisher` is an `ign-gui` plugin that reads the keyboard's keystrokes and sends them on a default topic `/keyboard/keypress`.
To use this plugin add the following code under the `<gui>` tag.

```xml
<!-- KeyPublisher plugin-->
<plugin filename="KeyPublisher" name="Key Publisher"/>
```

Let's try this plugin as follows:

* In one terminal type

>     `ign gazebo car_demo.sdf`

Reviewer: ```suggestion
    `ign gazebo building_robot.sdf`
```

![KeyPublisher](keypublisher_data.png)

We want to map these keystrokes into messages of type `Twist` and publish them to the `/cmd_vel` topic which our model listens to.
The `TriggeredPublisher` plugin will do this.

### Triggered Publisher

The `TriggeredPublisher` plugin publishes a user specified message on an output topic in response to an input message that matches user specified criteria.
Let's add the following code under the `<world>` tags:

```xml
<!-- Moving Forward-->
<plugin filename="libignition-gazebo-triggered-publisher-system.so"
        name="ignition::gazebo::systems::TriggeredPublisher">
    <input type="ignition.msgs.Int32" topic="/keyboard/keypress">
        <match field="data">16777235</match>
    </input>
    <output type="ignition.msgs.Twist" topic="/cmd_vel">
        linear: {x: 0.5}, angular: {z: 0.0}
    </output>
</plugin>
```

This code defines the `triggered-publisher` plugin.
It accepts messages of type `ignition.msgs.Int32` on the `/keyboard/keypress` topic and if the value in the `data` field matches `16777235`(Up arrow key) it outputs a `Twist` message on the `cmd_vel` topic with values `x: 0.5`, `z: 0.0`.

> Now launch `car_demo.sdf` and our robot should move forward as we press the Up arrow key &#8593;.

Reviewer: ```suggestion
Now launch `building_robot.sdf` and our robot should move forward as we press the Up arrow key &#8593;.
```

There is a demo explaining how the [Triggered Publisher](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo2/tutorials/triggered_publisher.md) works.

### Moving using arrow keys

To see what values are sent on the `/keyboard/keypress` topic when pressing the arrows we can use the `--echo` or `-e` option

* Run the model in one terminal:

>     `ign gazebo car_demo.sdf`

Reviewer: ```suggestion
    `ign gazebo building_robot.sdf`
```
